overlord/snapshotstate/backend/backend.go

```diff
@@ -378,7 +387,54 @@ func Save(ctx context.Context, id uint64, si *snap.Info, cfg map[string]interfac
 
 var isTesting = snapdenv.Testing()
 
-func addDirToZip(ctx context.Context, snapshot *client.Snapshot, w *zip.Writer, username string, entry, dir string) error {
+type addDirToZipOptions struct {
+	ExcludePaths []string `yaml:"exclude"`
+}
+
+func readSnapshotYaml(si *snap.Info, opts *addDirToZipOptions) error {
+	file, err := osOpen(filepath.Join(si.MountDir(), "meta", "snapshots.yaml"))
+	if err != nil {
+		if !os.IsNotExist(err) {
+			return err
+		} else {
+			return nil
+		}
+	}
+	defer file.Close()
+
+	if err := yaml.NewDecoder(file).Decode(opts); err != nil {
+		return fmt.Errorf("cannot read snapshot manifest: %v", err)
+	}
+
+	// Validate the exclude list; note that this is an *exclusion* list, so
+	// even if the manifest specified paths starting with ../ this would not
+	// cause tar to navigate into those directories and pose a security risk.
+	// Still, let's have a minimal validation on them being sensible.
+	validFirstComponents := []string{
+		"$SNAP_DATA", "$SNAP_COMMON", "$SNAP_USER_DATA", "$SNAP_USER_COMMON",
+	}
+	for _, excludePath := range opts.ExcludePaths {
+		firstSlash := strings.IndexRune(excludePath, '/')
```

Comment:
(nitpick) AFAICT this could be written more compact as `firstComponent := strings.SplitN(excludePath, "/", 2)[0]`:
```go
@@ -414,24 +414,18 @@ func readSnapshotYaml(si *snap.Info, opts *addDirToZipOptions) error {
                "$SNAP_DATA", "$SNAP_COMMON", "$SNAP_USER_DATA", "$SNAP_USER_COMMON",
        }
        for _, excludePath := range opts.ExcludePaths {
-               firstSlash := strings.IndexRune(excludePath, '/')
-               var firstComponent string
-               if firstSlash < 0 {
-                       firstComponent = excludePath
-               } else {
-                       firstComponent = excludePath[:firstSlash]
-               }
+               firstComponent := strings.SplitN(excludePath, "/", 2)[0]
                if !strutil.ListContains(validFirstComponents, firstComponent) {
...
```

Reply:
I did this now, hope I did not overlook something :)